lib/src/generate/build_impl.dart

```diff
@@ -294,3 +305,155 @@ class BuildImpl {
     }
   }
 }
+
+/// Extends [BuildImpl] to add the [runWatch] method, which watches all
+/// inputs for changes.
+class WatchImpl extends BuildImpl {
+  /// Injectable factory for creating directory watchers.
+  final DirectoryWatcherFactory _directoryWatcherFactory;
+
+  /// Delay to wait for more file watcher events.
+  final Duration _debounceDelay;
+
+  /// Whether or not we are currently watching and running builds.
+  bool _runningWatch = false;
+
+  /// Controller for the results stream.
+  StreamController<BuildResult> _resultStreamController;
+
+  /// All file listeners currently set up.
+  final _allListeners = <StreamSubscription>[];
+
+  /// A future that completes when the current build is done.
+  Future _currentBuild;
+
+  /// Whether or not another build is scheduled.
+  bool _nextBuildScheduled;
+
+  WatchImpl(
+      this._directoryWatcherFactory,
+      this._debounceDelay,
+      AssetCache cache,
+      AssetGraph assetGraph,
+      AssetReader reader,
+      AssetWriter writer,
+      PackageGraph packageGraph,
+      List<List<Phase>> phaseGroups)
+      : super(cache, assetGraph, reader, writer, packageGraph, phaseGroups);
+
+  /// Completes after the current build is done, and stops further builds from
+  /// happening.
+  Future terminate() async {
+    assert(_runningWatch == true);
+    _logger.info('Terminating watchers, no futher builds will be scheduled.');
+    _nextBuildScheduled = false;
+    _runningWatch = false;
+    for (var listener in _allListeners) {
+      await listener.cancel();
+    }
+    _allListeners.clear();
+    if (_currentBuild != null) {
+      _logger.info('Waiting for ongoing build to finish.');
+      await _currentBuild;
+    }
+    await _resultStreamController.close();
```

Comment:
This could be `null` if a user calls `terminate` prior to `runWatch` -- check above & throw a `StateError` if this happens.